nemo/collections/nlp/data/dialogue/data_processor/custom_data_processor.py

```diff
@@ -0,0 +1,210 @@
+# Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+from typing import List
+
+from tqdm import tqdm
+
+from nemo.collections.nlp.data.dialogue.data_processor.data_processor import DialogueDataProcessor
+from nemo.collections.nlp.data.dialogue.input_example.input_example import DialogueInputExample
+from nemo.utils import logging
+
+
+class DialogueCustomDataProcessor(DialogueDataProcessor):
+    """ Data processor for a custom dialog bot dataset """
+
+    def __init__(self, data_dir: str, tokenizer: object = None, cfg=None):
+        """ Constructs the dialog data processor """
+
+        self.data_dir = data_dir
+        self.cfg = cfg
+        self._task_name = self.cfg.task_name
+        self._tokenizer = tokenizer
+
+    def get_dialog_examples(self, dataset_split) -> List[object]:
+        """
+        Reads dialogue examples from self.data_dir/dataset_split/examples.json
+        and formats into SGD structure
+        """
+
+        dialogues_filepath = os.path.join(self.data_dir, dataset_split, "examples.json")
+        if not os.path.exists(dialogues_filepath):
+            raise ValueError(f"Dialogues file {dialogues_filepath} does not exist")
+
+        logging.info(f"Reading dialogue examples from {dialogues_filepath}")
+        with open(dialogues_filepath, "r") as f:
+            data = json.load(f)["data"]
+
+        logging.info(f"Processing dialogues...")
+        examples = []
+        for dialogue_idx, dialogue in enumerate(tqdm(data)):
+
+            # skip dialogues where system slots are absent
+            if not dialogue["Fulfillment"]:
+                continue
+
+            DialogueCustomDataProcessor.add_metadata_to_fulfillment_slots(dialogue)
+            dialogue_dict = DialogueCustomDataProcessor.create_dialogue_sgd_format_dict(dialogue, dialogue_idx)
+            if dialogue_dict:
+                curr_examples = DialogueCustomDataProcessor.create_examples_from_dialogue_dict(dialogue_dict)
+                examples.extend(curr_examples)
+
```

Comment:
The logic below is pretty complicated (mostly because SGD is rather complex). I think as a sanity check, you should either print out one or two examples or dump into a file all of the examples to see if they are what you are expecting.

Reply:
If you want to do the later, you can do json_examples = [example.data for example in examples] and then dump the json_examples into a json_file